records/2024-05-23.scalability.short-code-implementation.md

```diff
@@ -0,0 +1,97 @@
+Usage of a pool of codes/ dedicated long codes
+
+## Current Context and Assumptions
+
+1. **We assume that ALL Priority Messages will be sent using the Short Code**
+1. For the purposes of this ADR, assume SMSs are sent to a send_sms client call. Under the hood the send_sms call might SMS or Piinpoint.
+1. Currently, we have 3 priority queues - HIGH, MEDIUM, and LOW. Given a template, your message will reach within a certain SLO if sent through a certain queue
+1. A template can be of type PRIORITY, NORMAL, and BULK which ***currently*** maps directly to the HIGH, MEDIUM, and LOW queues. 
+1. The templates refer to both SMS and Email templates, but for shortcodes, this ADR focuses on SMS use cases
+
+# Implementation
+
+## Template Categorization
+
+When a client creates a template we want to be able to Categorize it into a certain type of template. We will start with the following Template Categories that were created through research on Notify's current templates.
+
+1. Status updates (eg. notice of change in status, the progress of a submission)
+1. Promotional call to action (eg. surveys, general apply now, learn more)
+1. Service-related requests (eg. submit additional documents, follow up to move a process forward) 
+1. Fulfillment with attachments - email only (eg. here’s your permit)
+Broadcast messages (eg. general information, not related to transactions such as COVID-19 information)
+1. Auto-reply (eg. no-reply messages, acknowledgments, response wait times)
+1. Verification message (eg. authentication codes, confirming an account change)
+1. Confirmation / Receipts (eg. record of transaction, approvals)
+1. Other
+
+### Determining Template Priority
+A new table will be added to the DB for TemplateCategories. The primary key from TemplateCategories will be a foreign key to the Templates Table. All TemplateCategories will be assigned a pre-computed priority.
+
+<!-- ![Potential ERD](https://raw.githubusercontent.com/cds-snc/notification-adr/records/diagrams/2024-05-23.scalability.short-code-implementation/ERD.jpg?raw=true "Potential ERD") -->
+
+![Potential ERD](../records/diagrams/2024-05-23.scalability.short-code-implementation/ERD.jpg?raw=true "Potential ERD")
+
+```
+process_type is (bulk, normal, priority)
+```
+
+Given the new TemplateCategories, each template will default to a process_type that is set in TemplateCategory. To create manual overrides (admins will be able to do so for a given template) - Admins will edit the edit the process_type on the Templates Table. For this to work, we have two options:
+1. Create a new column in Templates `new_process_type`
+2. Clear the current process_type in the Templates table to be empty.
+
+After the above is implemented,
+```
+Notification Priority = Template.new_process_type else TemplateCategory.process_type
+```
+For most Templates, the Template.process_type/new_process_type will be empty. Only if overridden by Admin, will that field have information.
```

Comment:
If the chosen category determines the process_type and we default it to `low/bulk` in the case of a newly added categories via the `Other` option, then when would `process_type` be empty?

Reply:
There is a process_type in TemplateCategories and process_type in Templates. 
1. If a client chooses a category, the process_type in Templates will be empty. 
1. If the client chooses "Other" - while we are determining what TemplateCategory the Template must have, it will default to "Low".
1. If an Admin wants to override the process_type for a given Template, that process_type will be overwritten in the Templates table.

Reply:
This makes sense to me since we can override the `process_type`.  I initially thought that the `process_type` for the template would be set with the one associated with the `TemplateCategory`.

Reply:
That could be a third option @whabanks where we could override all `Template.process_type` column to be initially set with the `TemplateCategory.process_type` value... I like though the blank option because it carries extra meaning that we did override the process type and a decision was specifically made for the associated template/service.